helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -152,69 +179,147 @@ private double getAssignmentNormalizedScore(AssignableNode node, AssignableRepli
+      Map<String, Integer> replicaHashCodeMap, float replica1Score, float replica2Score) {
+    Map<String, ResourceAssignment> _bestPossibleAssignment =
+        clusterModel.getContext().getBestPossibleAssignment();
+    Map<String, ResourceAssignment> _baselineAssignment =
+        clusterModel.getContext().getBaselineAssignment();
+
+    String resourceName1 = replica1.getResourceName();
+    String resourceName2 = replica2.getResourceName();
+
+    // 1. Sort according if the assignment exists in the best possible and/or baseline assignment
+    if (_bestPossibleAssignment.containsKey(resourceName1) != _bestPossibleAssignment
+        .containsKey(resourceName2)) {
+      // If the best possible assignment contains only one replica's assignment,
+      // prioritize the replica.
+      return _bestPossibleAssignment.containsKey(resourceName1);
+    }
+
+    if (_baselineAssignment.containsKey(resourceName1) != _baselineAssignment
+        .containsKey(resourceName2)) {
+      // If the baseline assignment contains only one replica's assignment, prioritize the replica.
+      return _baselineAssignment.containsKey(resourceName1);
+    }
+
+    // 2. Sort according to the state priority. Or the greedy algorithm will unnecessarily shuffle
+    // the states between replicas.
+    int statePriority1 = replica1.getStatePriority();
+    int statePriority2 = replica2.getStatePriority();
+    if (statePriority1 != statePriority2) {
+      // Note we shall prioritize the replica with a higher state priority,
+      // the smaller priority number means higher priority.
+      return statePriority1 - statePriority2 < 0;
+    }
+
+    // 3. Sort according to the replica impact based on the weight.
+    // So the greedy algorithm will place the more impactful replicas first.
+    if (Math.abs(replica2Score - replica1Score) > 0.001f) {
+      return replica1Score > replica2Score;
+    }
+
+    // 4. Sort according to the resource/partition name.
+    // If none of the above conditions is making a difference, try to randomize the replicas
+    // order.
+    // Otherwise, the same replicas might always be moved in each rebalancing. This is because
+    // their placement calculating will always happen at the critical moment while the cluster is
+    // almost close to the expected utilization.
+    //
+    // Note that to ensure the algorithm is deterministic with the same inputs, do not use
+    // Random functions here. Use hashcode based on the cluster topology information to get
+    // a controlled randomized order is good enough.
+    Integer replicaHash1 = replicaHashCodeMap.get(replica1.toString());
+    Integer replicaHash2 = replicaHashCodeMap.get(replica2.toString());
+    if (!replicaHash1.equals(replicaHash2)) {
+      return replicaHash1.compareTo(replicaHash2) < 0;
+    } else {
+      // In case of hash collision, return order according to the name.
+      return replica1.toString().compareTo(replica2.toString()) < 0;
+    }
```

Comment:
NIT: I suppose you could also have pre-sorted the list of replicas with the other 3 conditions, store the sorted list outside, then just compare the scores during `getNextAssignableReplica`. It doesn't provide any performance boost but it looks cleaner IMO.

Reply:
As discussed with @xyuanlu , pre-sorting will have additional n(log n) complexity, since the usage here is only "find max", not sorting. Feel free to discard this comment.